contracts/ibc-reflect/src/contract.rs

```diff
@@ -0,0 +1,205 @@
+};
+
+use crate::msg::{
+    AcknowledgementMsg, HandleMsg, InitMsg, PacketMsg, ReflectHandleMsg, ReflectInitMsg,
+};
+use crate::state::{accounts, config, Config};
+
+const IBC_VERSION: &str = "ibc-reflect";
+
+#[entry_point]
+pub fn init(deps: DepsMut, _env: Env, _info: MessageInfo, msg: InitMsg) -> StdResult<InitResponse> {
+    // we store the reflect_id for creating accounts later
+    let cfg = Config {
+        reflect_code_id: msg.reflect_code_id,
+    };
+    config(deps.storage).save(&cfg)?;
+
+    Ok(InitResponse::default())
+}
+
+#[entry_point]
+pub fn handle(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: HandleMsg,
+) -> StdResult<HandleResponse> {
+    match msg {
+        HandleMsg::InitCallback { id, contract_addr } => {
+            handle_init_callback(deps, info, id, contract_addr)
+        }
+    }
+}
+
+pub fn handle_init_callback(
+    deps: DepsMut,
+    info: MessageInfo,
+    id: String,
+    contract_addr: HumanAddr,
+) -> StdResult<HandleResponse> {
+    // sanity check - the caller is registering itself
+    if info.sender != contract_addr {
+        return Err(StdError::generic_err("Must register self on callback"));
+    }
+
+    // store id -> contract_addr if it is empty
+    // id comes from: `let chan_id = msg.endpoint.channel_id;` in `ibc_channel_connect`
+    accounts(deps.storage).update(id.as_bytes(), |val| -> StdResult<_> {
+        match val {
+            Some(_) => Err(StdError::generic_err(
+                "Cannot register over an existing channel",
+            )),
+            None => Ok(contract_addr),
+        }
+    })?;
+
+    Ok(HandleResponse::default())
+}
+
+#[entry_point]
+/// enforces ordering and versioing constraints
+pub fn ibc_channel_open(_deps: DepsMut, _env: Env, channel: IbcChannel) -> StdResult<()> {
+    if channel.order != IbcOrder::Ordered {
+        return Err(StdError::generic_err("Only supports ordered channels"));
+    }
+    if channel.version.as_str() != IBC_VERSION {
+        return Err(StdError::generic_err(format!(
+            "Must set version to `{}`",
+            IBC_VERSION
+        )));
+    }
+
+    Ok(())
+}
+
+#[entry_point]
+/// once it's established, we create the reflect contract
+pub fn ibc_channel_connect(
+    deps: DepsMut,
+    _env: Env,
+    channel: IbcChannel,
+) -> StdResult<IbcBasicResponse> {
+    let cfg = config(deps.storage).load()?;
+    let chan_id = channel.endpoint.channel_id;
+    let label = format!("ibc-reflect-{}", &chan_id);
+
+    let payload = to_binary(&ReflectInitMsg {
+        callback_id: Some(chan_id),
+    })?;
+    let msg = WasmMsg::Instantiate {
+        code_id: cfg.reflect_code_id,
+        msg: payload,
+        send: vec![],
+        label: Some(label),
+    };
+
+    Ok(IbcBasicResponse {
+        messages: vec![msg.into()],
+        attributes: vec![],
+    })
+}
+
+#[entry_point]
+/// we do nothing
+/// TODO: remove the account from the lookup?
```

Comment:
I think this would never actually be called.

The flow I have is "to-be-defined-sender-contract on chain A" sends a packet to "ibc-reflect on chain B", which dispatches it.

`ibc-reflect` never sends a packet itself, thus should never get these two callbacks (we could error here I guess?)